nabs/event.py

```diff
@@ -0,0 +1,109 @@
+from collections import defaultdict
+from functools import partial
+from threading import Event, Thread
+import math
+import time
+
+from ophyd.status import Status
+
+
+class EventBuilder:
+    def __init__(self, signals, rate, timeout=10.0, duplicates='closest'):
+        self.signals = signals
+        self.rate = rate
+        self.timeout = timeout
+        self.duplicates = duplicates
+        self.cbid = {}
+        self.start_ts = 0
+        self.name = 'event_builder'
+        self._clear_events()
+
+    def add(self, name, *, value, timestamp, **kwargs):
+        if timestamp > self.start_ts:
+            bin_ts = self._nearest_ts(timestamp)
+            delta = timestamp - bin_ts
+            event = self.bins[bin_ts]
+            event[name].append({'value': value,
+                                'timestamp': timestamp,
+                                'delta': delta})
+            if len(event) == len(self.signals):
+                self._has_evt.set()
+
+    def _nearest_ts(self, timestamp):
+        expanded = timestamp * self.rate
+        floor = math.floor(expanded) / self.rate
+        ceil = math.ceil(expanded) / self.rate
+        if timestamp - floor < ceil - timestamp:
+            return floor
+        else:
+            return ceil
+
+    def _monitor_events(self):
+        if not self.cbid:
+            self.start_ts = time.time()
+            for sig in self.signals:
+                cbid = sig.subscribe(partial(self.add, sig.name))
+                self.cbid[sig.name] = cbid
+
+    def _stop_events(self):
+        if self.cbid:
+            for sig in self.signals:
+                sig.unsubscribe(self.cbid[sig.name])
+        self.cbid = {}
+
+    def _clear_events(self):
+        self._stop_events()
+        self._has_evt = Event()
+        self.bins = defaultdict(partial(defaultdict, list))
```

Comment:
This line confuses me. I question whether it is cleaner than just having a `get(key, default)` call...

Reply:
I agree that this line is confusing, but the alternative is also bad.
```
event = self.bins.get(ts, {})
data = event.get(name, [])
data.append(<datagram>)
event[name] = data
self.bins[ts] = event
```
The `defaultdict` lets you skip the reassigning part (e.g. omit the last two lines). If you think this is ok I can look into doing the reassigning everywhere, it might only be in 2 or 3 places.

Reply:
Eh. Think its fine.